ros/src/tl_detector/dummy_detector.py

```diff
@@ -0,0 +1,192 @@
+            return
+
+        # Locate the next upcoming red traffic light stop line waypoint index
+        closest_stop_index = len(self.waypoints) - 1
+        for light in self.lights:
+            # Green light is not an obstacle!
+            if light.state == TrafficLight.RED:
+                
+                # Get the stop line from the light
+                light_x = light.pose.pose.position.x
+                light_y = light.pose.pose.position.y
+                stop_line = self.light_stops[(light_x, light_y)]
+                rospy.loginfo("found red @ %s",
+                              self.get_closest_waypoint(light_x, light_y))
+
+                # Get the waypoint index closest to the stop line
+                stop_line_x = stop_line[0]
+                stop_line_y = stop_line[1]
+                stop_wp_index = self.get_closest_waypoint(stop_line_x,
+                                                          stop_line_y)
+      
+                if  stop_wp_index > car_wp_index and \
+                    stop_wp_index < closest_stop_index:
+                    closest_stop_index = stop_wp_index
+
+        rospy.loginfo("closest stop @ %s", closest_stop_index)
+        
+        # Publish the result
+        self.upcoming_red_light_pub.publish(closest_stop_index)
+
+    #---------------------------------------------------------------------------
+    # Methods
+    #---------------------------------------------------------------------------
+
+    def correlate_lights_and_stop_positions(self):
+        """
+        Assign the closest stop line position to each of the traffic lights.
+        The operation is supposed to be done only once
+        """
+        for light in self.lights:
+            # Reset the minimum distance and the index we search for
+            min_dist = float('inf')
+            matching_index = 0
+            
+            for i, stop_line_position in enumerate(self.stop_line_positions):
+                # Calculate the Euclidean distance
+                dx = light.pose.pose.position.x - stop_line_position[0]
+                dy = light.pose.pose.position.y - stop_line_position[1]
+                dist = pow(dx, 2) + pow(dy, 2)
+                
+                # Update the minimum distance and matching index
+                if dist < min_dist:
+                    min_dist = dist
+                    matching_index = i
+        
+            # Correlate each light position (x, y) with the closest stop line
+            x = light.pose.pose.position.x
+            y = light.pose.pose.position.y
+            self.light_stops[(x, y)] = self.stop_line_positions[matching_index]
+
+            rospy.loginfo("light = (%s, %s) : stop = (%s, %s)",
+                          x, y, self.stop_line_positions[matching_index][0],
+                          self.stop_line_positions[matching_index][1])
+
+    def get_closest_waypoint(self, x, y):
```

Comment:
get_closeset_waypoint is currently O(n) and this is called at every traffic callback. On my machine this took over 8ms. This is mostly critical for live tracking. But even with simulated tracking we see discernible improvement when using a tree to search.

Reply:
Shall we merge this and then call your tree search function from the dummy?